interfaces/builtin/home.go

```diff
@@ -63,14 +76,56 @@ owner /run/user/[0-9]*/gvfs/{,**} r,
 owner /run/user/[0-9]*/gvfs/*/**  w,
 `
 
+const homeConnectedPlugAppArmorWithAllRead = `
+# Allow non-owner read to non-hidden and non-snap files and directories
+@{HOME}/               r,
+@{HOME}/[^s.]**        r,
+@{HOME}/s[^n]**        r,
+@{HOME}/sn[^a]**       r,
+@{HOME}/sna[^p]**      r,
+@{HOME}/snap[^/]**     r,
+@{HOME}/{s,sn,sna}{,/} r,
+`
+
+type homeInterface struct {
+	commonInterface
+}
+
+func (iface *homeInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	// It's fine if 'read' isn't specified, but if it is, it needs to be
+	// either 'all' or 'owner'
+	if r, ok := plug.Attrs["read"]; ok {
+		_, ok = r.(string)
```

Comment:
Perhaps this is correct but it reads as a bit off.

r is an interface (probably `interface{}`) so you need a type assertion to see that it is a string.

But after the test r is still `interface{}` because you didn't assign the type-safe `r` anywhere (in pseudo-code: `rs, ok := r.(string)`. I'm not sure what go does in the case that ok is true (so `r` is a string) but the comparisons use an interface on the lvalue and a string literal on the rvalue.

Reply:
AFAIK Go does the right thing (although it does look 'different' from the usual pattern). Interfaces are comparable and equal if both have the same dynamic type and dynamic values are equal.

Reply:
The logic is safe, but there's another gotcha here in terms of the actual intention behind the logic: this check ignores incorrect values that have the wrong type.

We can both fix that and make the logic simpler by folding both conditionals into:

if r, ok := plug.Attrs["read"]; ok && r != "all" {
    ... error ...
}

Reply:
Sorry, my point about the mishandling of the wrong type is incorrect. The simplification point remains.